proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,664 @@
+
+```swift
+await withTaskExecutor(someExecutor) { 
+ // guaranteed to be executing on someExecutor
+}
+```
+
+Once set, the effect of an executor preference is such that a nonisolated func instead of immediately hopping to the global pool, it may hop to the preferred executor, e.g.:
+
+```swift
+nonisolated func tryMe() async { 
+ // ...
+}
+
+let preferredExecutor: SomeConcreteExecutor = ...
+Task(on: preferredExecutor) { 
+  preferredExecutor.assertIsolated()
+  await tryMe() // tryMe body would execute on 'preferredExecutor'
+}
+
+ await tryMe() // tryMe body would execute on 'default global concurrent executor'
+```
+
+### Task executor preference inheritance in Structured Concurrency
+
+Task executor preference is inherited by child tasks and actors which do not declare an explicit executor (so-called "default actors"), and is *not* inherited by un-structured tasks. 
+
+Specifically:
+
+* **Do** inherit task executor preference
+    * TaskGroup’s `addTask()`, unless overridden with explicit parameter
+    * `async let`
+    * methods on actors which which do not declare an explicit `unownedExecutor` requirement
+* **Do not** inherit task executor preference
+    * Unstructured tasks: `Task {}` and `Task.detached {}`
+    * methods on actors which **do** declare an explicit `unownedExecutor` (including e.g. the `MainActor`)
+
+This also means that an entire tree can be made to execute their nonisolated work on a specific executor, just by means of setting the preference on the top-level task.
+
+### Task executor preference and async let
+
+Since `async let` are the simplest form of structured concurrency, they dot not offer much in the way of customization.
+
+An async currently always executes on the global concurrent executor, and with the inclusion of this proposal, it does take into account task executor preference. In other words, if an executor preference is set, it will be used by async let to enqueue its underlying task:
+
+```swift
+func test(_ someExecutor: any Executor) -> Int {
+  someExecutor.assertIsolated()
+  return 42
+}
+
+await withTaskExecutor(someExecutor) { 
+  async let value = test(someExecutor) // executes on 'someExecutor'
+  // ... 
+  await value
+}
+```
+
+### Task executor preference and TaskGroups
+
+A `TaskGroup` and its various friends (`ThrowingTaskGroup`, `DiscardingTaskGroup`, ...) are the most powerful, but also most explicit and verbose API for structured concurrency. A group allows creating multiple child tasks using the `addTask` method, and always awaits all child tasks to complete before returning.
+
+This proposal adds overloads to the `addTask` method, which changes the executor the child tasks will be enqueued on:
+
+```swift
+extension (Discarding)(Throwing)TaskGroup { 
+  mutating func addTask(
+    on executor:  (any Executor)?,  // 'nil' means the global pool
```

Comment:
Just so I get it right. We introduce the new `withExecutor` method and all existing code that is called in such a scope would automatically inherit. Now developers might go around and check if there is code that should not inherit and then add `on: nil`, right?

Could we instead spell it `on: .global`? I think the `nil` is a bit confusing

Reply:
We can discuss that in evolution yeah